src/proto/h2/mod.rs

```diff
@@ -243,32 +241,152 @@ impl<B: Buf> SendStreamExt for SendStream<SendBuf<B>> {
 
     fn send_eos_frame(&mut self) -> crate::Result<()> {
         trace!("send body eos");
-        self.send_data(SendBuf(None), true)
+        self.send_data(SendBuf::None, true)
             .map_err(crate::Error::new_body_write)
     }
 }
 
-struct SendBuf<B>(Option<B>);
+enum SendBuf<B> {
+    Buf(B),
+    Cursor(Cursor<Box<[u8]>>),
```

Comment:
Whatcha think of this being `Bytes`, and `H2Upgraded::write` pushing into a `BytesMut`? I can't say if it's actually better, it'd only be better in these situations:

- The `BytesMut` reserves more space and the `split_off` means the other space is used without allocating.
- The sliced off `Bytes` are flushed on the connection and dropped before another `poll_write`, which means the `BytesMut` would know it could reuse its buffer instead of allocating a new one.

Not a strong desire here, just curious what you think.

Reply:
I'm sorry I don't understand either of your bullet points 😅

What would the `BytesMut` do, and when would we actually send the buffer to the h2 stream? Do you mean that `poll_write` would never send a partial buffer anymore? Wouldn't that be bad for flow control?

Reply:
It was more just a passing thought, but I don't have data that in this instance it would be an improvement. Maybe something to try out separately and see if there is actually any performance improvement, or if I'm spouting nonsense.